JBPM-8195: GraphQL as a modern access to jBPM process engine

domhanak opened the pull request:
@mswiderski Hello :) Could you please take a look when you have a second? 

I still need to split client and integration test into respective modules.
I am hacking dependency management of anltr4-runtime for now. My goal is 
to upgrade it to 4.7.2 - There is PR in community to test the upgrade.
I will be adding more operations in the meantime.


Depends on https://github.com/domhanak/droolsjbpm-build-bootstrap/pull/1

domhanak committed:
JBPM-8195: Setup kie-server-graphql-jbpm module skeleton

Add pom.xml files with basic data
Add gitignore files
Add kie-server-graphql module into kie-server-remote module declaration

domhanak committed:
JBPM-8195: Add initial schema declaration

Add schema with basic structure.
Add initial jbpm type - ProcessDefinition & ProcessInstance.
Add descriptions to each type.
Add basic mutations for instance.
Add basic queries for definitions and instances.

domhanak committed:
JBPM-8195: Add serviceProvider for graphql extension

domhanak committed:
JBPM-8195: Create Resource object and components service

domhanak committed:
Add service declaration to META-INF/services

domhanak committed:
JBPM-8195: Add necessary dependencies

JBPM service dependencies.
GraphQL dependencies.
Test dependencies.

domhanak committed:
Add root Java classes for GraphQL mutations and queries

domhanak committed:
Add implementation for queries and mutations defined in schema

Add InstanceMutation resolver.
Add InstanceQuery resolver.
Add Definitionquery resolver.
Add filter objects for tasks and processes.

domhanak committed:
Update the schema to reflect resovler changes

domhanak committed:
Add unit test for database connectors

domhanak committed:
Add constants and exceptions for graphql extension

domhanak committed:
JBPM-8195: Add kie-server-client module

domhanak committed:
JBPM-8195: Create graphql client for jbpm

domhanak committed:
JBPM-8195: Adds response object for client

domhanak committed:
JBPM-8195: Add skeletons for unit testing reslovers

domhanak committed:
Add graphql extension services module

domhanak committed:
JBPM-8195: Add missing gitignore file to client

domhanak committed:
JBPM-8195: Fix name of client module

domhanak committed:
JBPM-8195: Add module for integration-tests

domhanak committed:
JBPM-8195: Add integration tests for mutation and query

domhanak committed:
JBPM-8195: Declare modules in parent kie-server-graphql module

domhanak committed:
JBPM-8195: Add ClientBuilder declaration file

domhanak committed:
JBPM-8195: Rename instance mutations to match schema

domhanak committed:
JBPM-8195: Fix capablity for graphql client

domhanak committed:
JBPM-8195: Update integration-tests to reflect schema changes

domhanak committed:
JBPM-8195: Update kie-server assemblies and add graphql jbpm dependencies

domhanak committed:
JBPM-8195: Fix client

domhanak committed:
JBPM-8195: Fix illegal transitive errors and modify IT test to work

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm-services/src/main/java/org/kie/server/remote/graphql/jbpm/JbpmGraphQLKieServerExtension.java:
```diff
@@ -0,0 +1,171 @@
+public class JbpmGraphQLKieServerExtension implements KieServerExtension {
+    public static final String EXTENSION_NAME = "JBPM GraphQL";
+
+    private static final Boolean JBPM_DISABLED = Boolean.parseBoolean(System.getProperty(KieServerConstants.KIE_JBPM_SERVER_EXT_DISABLED, "false"));
```
shouldn't you have dedicated property for this extension? Obviously it should be disabled when the jBPM extension is disabled - similar to jBPM UI or case mgmt extension

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm-services/src/main/java/org/kie/server/remote/graphql/jbpm/JbpmGraphQLKieServerExtension.java:
```diff
@@ -0,0 +1,171 @@
+
+    @Override
+    public Integer getStartOrder() {
+        // To be started after Drools
```
I guess it needs to also start after jBPM as it relies on its services

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm-tests/src/test/filtered-resources/kjars-sources/definition-project/src/main/java/org.jbpm.data/CustomCommand.java:
```diff
@@ -0,0 +1,32 @@
+import org.kie.api.executor.CommandContext;
+import org.kie.api.executor.ExecutionResults;
+
+public class CustomCommand implements Command{
```
is this used?

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm-tests/src/test/java/org/kie/server/integrationtests/graphql/jbpm/JbpmGraphQLExtensionIntegrationTest.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Override
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses) throws Exception {
+       //extraClasses.put(PERSON_CLASS_NAME, Class.forName(PERSON_CLASS_NAME, true, kieContainer.getClassLoader()));
```
removed if not needed and same for the Person class from Kjar if not used

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/main/java/org/kie/server/remote/graphql/jbpm/JbpmGraphQLResource.java:
```diff
@@ -0,0 +1,131 @@
+    public Map<String, Object> graphqlGET(@Context HttpHeaders headers,
+                                          @QueryParam("query") String query,
+                                          @QueryParam("operationName") String operationName,
+                                          @QueryParam("variables") String variablesJson) throws IOException {
```
how big the variables can be? maybe it would be better to use POST for it?

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/main/java/org/kie/server/remote/graphql/jbpm/JbpmGraphQLServiceProvider.java:
```diff
@@ -0,0 +1,45 @@
+ * Holds every service JBPM GraphQL extension needs
+ * These services are final and once set they can't be altered.
+ */
+public final class JbpmGraphQLServiceProvider {
```
jBPM Service api expose ServiceRegistry that already holds references to the jBPM services so might be an alternative to use instead.

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/main/java/org/kie/server/remote/graphql/jbpm/mutation/InstanceMutation.java:
```diff
@@ -0,0 +1,159 @@
+                                                DataFetchingEnvironment environment) {
+        boolean withVars = environment.getSelectionSet().contains(VARIABLES);
+        if (environment.getArgument(CORRELATION_KEY) != null
+                && environment.getArgument(PROCESS_VARIABLES) != null) {
```
do we need all these checks? sounds bit too much like api should be able to deal with null variables or null correlation key...

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/test/java/org/kie/server/remote/graphql/jbpm/mutation/InstanceMutationTest.java:
```diff
@@ -0,0 +1,5 @@
+package org.kie.server.remote.graphql.jbpm.mutation;
+
+public class InstanceMutationTest {
```
are these place holders for now or?

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/main/java/org/kie/server/remote/graphql/jbpm/repository/InstanceRepository.java:
```diff
@@ -0,0 +1,584 @@
+     * @return List started processInstances
+     */
+    public List<ProcessInstance> startProcesses(String id, String containerId, String correlationKey,
+                                                Map<String, Object> processVariables, int batchSize, boolean withVars) {
```
an idea that we could have the batch size calculated based on variables given, meaning that there would be a list of maps and for each item of the list we start new process instance, wdyt?

mswiderski commented on kie-server-parent/kie-server-remote/kie-server-graphql/kie-server-graphql-jbpm/src/main/resources/jbpm-schema.graphql:
```diff
@@ -0,0 +1,270 @@
+  """
+  Gets a list of ProcessDefinition based no arguments.
+  """
+  getProcessDefinitions(processDefinitionId: String, containerId: String): [ProcessDefinition]
```
so the difference between this and the processDefinition is that it can either take process definition by id or by container but not both right?

mswiderski reviewed (commented):
looks really nice!

Left few comments/questions to start with something.

Would you mind to bring others into the review?

mswiderski commented:
oh and one important bit, would you mind to create a readme file describgin few of the uses cases with sample curl commands?

domhanak commented:
Thanks, I will finish splitting client and tests to a existing modules, add more unit tests so that there are no placeholders :) Add the README and then a PR into community will follow, where more people can review (but feel free to send this bit to any1 who is interested) :)
Thanks for the feedback.

domhanak committed:
JBPM-8195: Add unit tests for InstanceQuery

domhanak committed:
JBPM-8195: Improve documentation of queries and mutation

Adds JavaDoc.
Fixes some queries and makes then better.
Add documentation for schema.

domhanak committed:
JBPM-8195: Fix new extension name

domhanak committed:
JBPM-8195: Add missing tests and fix current ones

domhanak committed:
JBPM-8195: Minor refactoring

domhanak committed:
JBPM-8195: Add missing constant

domhanak committed:
JBPM-8195: Refactor integration test

domhanak committed:
Add compile dependency on jbpm grapql service module

domhanak committed:
JBPM-8195: Add GraphQL extension name to KieserverConstants

domhanak committed:
Fix parent module of graphql extension

domhanak committed:
JBPM-8195: Introduce input types for mutations with minor refactor.

Add input types for signal, start and abort ops.
Fix tests and remove startProcess query.

domhanak committed:
JBPM-8195: Fix signal and abort not returning variables

Fixes queries not returning variables.
Minor refactoring and test fixes.

domhanak committed:
JBPM-8195: Improve schema documentation

Fix some arguments being too long.
Add more comments for missing types.

domhanak committed:
JBPM-8195: Add documentation for the schema and basic intro README

Add README.md with basic introduction to the API
Add jbpm-schema-documentation.md file.

domhanak committed:
JBPM-8195: Fix README.md and schema files

domhanak committed:
JBPM-8195: Fix jbpm-schema-documentation.md issues

Minor fixes for strings

domhanak committed:
JBPM-8195: Refactor tests and cleanup the code

domhanak committed:
JBPM-8195: Cleanup unused imports

domhanak commented:
Closing to follow up in main community repo.
